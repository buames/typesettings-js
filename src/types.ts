// tslint:disable max-line-length
import {
  FontFamilyProperty,
  FontSizeProperty,
  FontStyleProperty,
  FontWeightProperty,
  LineHeightProperty,
  LetterSpacingProperty,
  TextTransformProperty
} from 'csstype';

export enum FontSourceFormats {
  locals = 'locals',
  woff2 = 'woff2',
  woff = 'woff',
  ttf = 'ttf',
  otf = 'opentype',
  eot = 'embedded-opentype'
}

export enum LetterCasing {
  normalcase = 'normalcase',
  uppercase = 'uppercase',
  lowercase = 'lowercase'
}

export type StyledValue = string | number | undefined;

export type StyledObject = { [property: string]: StyledValue };

export type FontSources = { [K in FontSourceFormats]?: string | string[] | NodeRequireFunction | false };

export type TypesettingResults = { [size: string]: { [weight: string]: StyledFont } };

export interface TypesettingOptions {
  [k: string]: any;
  cssFn?: (styles: StyledObject | string) => any;
  fontStyles?: StyledObject;
  fontFaceStyles?: StyledObject;
}

export interface Typesettings {
  family: FontFamilyProperty;
  variants: FontVariant[];
  fallbacks?: FontFamilyProperty[];
}

export interface FontVariant {
  fontStyle: FontStyleProperty;
  fontWeight: FontWeightProperty;
  sources: FontSources;
  normalcase?: FontSetting[];
  uppercase?: FontSetting[];
  lowercase?: FontSetting[];
}

export interface FontSetting {
  fontSize: FontSizeProperty<StyledValue>;
  letterSpacing?: LetterSpacingProperty<StyledValue> | null;
  lineHeight?: LineHeightProperty<StyledValue> | null;
}

export interface StyledFont extends StyledObject {
  fontFamily: FontFamilyProperty;
  fontStyle?: FontStyleProperty;
  fontSize?: FontSizeProperty<StyledValue>;
  fontWeight?: FontWeightProperty;
  letterSpacing?: LetterSpacingProperty<StyledValue>;
  lineHeight?: LineHeightProperty<StyledValue>;
  textTransform?: TextTransformProperty;
}
