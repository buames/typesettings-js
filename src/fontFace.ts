import { Typesettings, FontSourceFormats, TypesettingOptions } from './types';
import { getFamilyName } from './utils';

export const generateFontFace = (
  typesettings: Typesettings,
  options: TypesettingOptions = { }
): string => {
  const { family, variants } = typesettings;
  const declaration = variants.map(({ sources, fontStyle, fontWeight }) => {
    if (!sources || (sources && Object.keys(sources).length === 0)) {
      throw Error('Missing font file sources');
    }

    const srcs = Object.keys(sources).map(key => (
      sources[key] && (Array.isArray(sources[key])
        ? sources[key].map((n: string) => (`local('${ n }')`))
        : `url(${ sources[key] }) format('${ FontSourceFormats[key] }')`)
    )).filter(Boolean);

    const face = [
      `font-family: ${ getFamilyName(family) }`,
      fontStyle && `font-style: ${ fontStyle }`,
      fontWeight && `font-weight: ${ fontWeight }`,
      srcs && `src: ${ srcs.join(', ') };`,
      options.fontFaceStyles
    ].filter(Boolean);

    return `@font-face { ${ face.join('; ') } }`;
  });

  const fontFace = declaration.join(' ');
  return options.cssFn ? options.cssFn(fontFace) : fontFace;
};
