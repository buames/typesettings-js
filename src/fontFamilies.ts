import { FontFamilyProperty } from 'csstype';

const genericFamilyKeywords = [
  'initial',
  'inherit',
  'unset',
  'serif',
  'sans-serif',
  'cursive',
  'fantasy',
  'monospace',
  'system-ui',
];

/**
 * Checks whether or not a font family name is a generic font family keyword
 *
 * When a generic family keyword keyword is wrapped in quotes, the browser will not interpret
 * them as keywords. Instead it looks for a font by that name (e.g. "sans-serif" font),
 * which is likley not what you actually want.
 *
 * @see https://mathiasbynens.be/notes/unquoted-font-family
 */
const isKeyword = (family: FontFamilyProperty) => {
  return genericFamilyKeywords.includes(family.toLowerCase());
};

/**
 * Checks whether or not a font family name is a vendor prefixed system font
 *
 * Quotes should never be used around vendor prefixed system fonts since the browser
 * will not interpret them as keywords
 */
const isSystemFont = (family: FontFamilyProperty) => {
  return family.startsWith('-apple-') || family === 'BlinkMacSystemFont';
};

/**
 * Checks whether or not a font family name should be wrapped in quotes
 *
 * It is recommended to quote font family names that contain white space, digits, or
 * punctuation characters other than hyphens.
 *
 * @see https://www.w3.org/TR/CSS2/fonts.html#font-family-prop
 */
const isCharactersOnly = (family: FontFamilyProperty) => {
  return /^[-a-zA-Z]+$/.test(family);
};

/**
 * Normalizes quotations for a given font family name or string of font family names.
 */
export const normalizeFontFamily = (value: FontFamilyProperty) => {
  return value
    .split(',')
    .map((raw: FontFamilyProperty) => {
      // Clean the family name, includes its quotes
      const family = raw.trim().replace(/^['"]|['"]$/g, '');

      if (
        isKeyword(family) ||
        isSystemFont(family) ||
        isCharactersOnly(family)
      ) {
        return family;
      }

      return `'${family}'`;
    }, [])
    .join(', ');
};

/**
 * Returns the font stack from a given family and array of fallbacks
 * */
export const getFontStack = (
  family: FontFamilyProperty,
  fallbacks?: FontFamilyProperty[],
) => {
  const stack = [
    normalizeFontFamily(family),
    fallbacks?.map(normalizeFontFamily).join(', '),
  ];

  return stack.filter(Boolean).join(', ');
};
