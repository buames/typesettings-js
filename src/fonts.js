/* eslint-disable object-curly-newline, no-param-reassign */
const merge = require('deepmerge')

/*
  Returns a weight property labelâ€š prefixed with 'n' for normal or 'i' for italics.
  For example, n700 equals 'normal' with a weight of '700'.
*/
const getStyleLabel = (fontStyle, fontWeight) => {
  switch (fontStyle) {
    case 'normal': return `n${ fontWeight }`
    case 'italics': return `i${ fontWeight }`
    default: return ''
  }
}

/*
  Returns a property label to append to the style property
  label depending on the lettercasing type
*/
const getTransformLabel = (lettercasing) => {
  switch (lettercasing) {
    case 'uppercase': return '_caps'
    case 'lowercase': return '_lower'
    case 'normalcase': return ''
    default: return ''
  }
}

/*
  Generates a map of typesettings. We do not return emotionjs classes
  as it would not work with media queries properly.
*/
const generate = (variant, options) => {
  const { casing, styles, family, fallback } = options
  const sizes = variant[casing]

  if (!sizes) {
    return { }
  }

  return Object.keys(sizes).reduce((accum, size) => {
    const { characterSpacing, lineHeight } = sizes[size]
    const styleLabel = getStyleLabel(variant.fontStyle, variant.fontWeight)
    const transformLabel = getTransformLabel(casing)

    accum[`s${ size }`] = { } // font-size property is prefixed with 's'
    accum[`s${ size }`][`${ styleLabel }${ transformLabel }`] = {
      fontFamily: fallback ? `${ family }, ${ fallback }` : family,
      fontSize: `${ size }px`,
      fontStyle: variant.fontStyle,
      fontWeight: variant.fontWeight,
      letterSpacing: `${ characterSpacing ? `${ characterSpacing.toFixed(2) }px` : 'initial' }`,
      lineHeight: `${ lineHeight ? `${ lineHeight }px` : 'normal' }`,
      textTransform: `${ casing === 'normalcase' ? 'none' : casing }`,
      ...styles || { }
    }

    return accum
  }, { })
}

const generateFonts = (typesettings, styles) => {
  if (typeof typesettings !== 'object') {
    throw new Error('Your typesettings must be an object.')
  }

  const { family, fallback, ...variants } = typesettings

  // Loop over the typeface variants and create each variant's
  // typesettings for each of their lettingcasing
  const settings = Object.values(variants).map((variant) => {
    const shared = { styles, family, fallback }
    const normal = generate(variant, { casing: 'normalcase', ...shared })
    const upper = generate(variant, { casing: 'uppercase', ...shared })
    const lower = generate(variant, { casing: 'lowercase', ...shared })
    return merge.all([ normal, upper, lower ])
  })

  return merge.all(settings)
}

export default generateFonts
