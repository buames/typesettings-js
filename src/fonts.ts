import {
  LetterCasing,
  StyledValue,
  Typesettings,
  TypesettingFontsOptions,
  TypesettingResults,
  FontVariant,
  FontSetting
} from './types';

import {
  getFontStack,
  parseSize
} from './utils';

/*
  Returns a pixel value or the raw css value
*/
const px = (n: StyledValue) => (
  typeof n === 'number' && n !== 0 ? `${ n }px` : n
);

/*
  Returns a weight property labelâ€š prefixed with 'n' for normal, 'i' for italics,
  and 'o' for oblique. For example, n700 equals 'normal' with a weight of '700'.
  However, if the weight is a string (ie 'bold'), this returns the prefix + a capitalized
  weight. For example, nBold.
*/
const getStyleLabel = ({ fontStyle, fontWeight }: FontVariant) => (
  `${ fontStyle.charAt(0) }${ typeof fontWeight === 'string'
    ? `${ fontWeight.charAt(0).toUpperCase() }${ fontWeight.slice(1) }`
    : fontWeight }`
);

/*
  Returns a property label to append to the style property
  label depending on the lettercasing type
*/
const getTransformLabel = (casing: LetterCasing | string) => (
  (casing === LetterCasing.uppercase) ? '_caps'
    : (casing === LetterCasing.lowercase) ? '_lower'
    : ''
);

/*
  Generates a map of typesettings
*/
export const generateFonts = (
  typesettings: Typesettings,
  options: TypesettingFontsOptions = { }
): TypesettingResults => {
  const { family, fallbacks, variants } = typesettings;
  const fontFamily = getFontStack(family, fallbacks);

  const styles: { } = variants.reduce((acc, variant) => {
    const { fontStyle, fontWeight, sources, ...casings } = variant;
    const styleLabel = getStyleLabel(variant);

    // Loop over each casing (ie. normalcase, uppercase, lowercase)
    Object.keys(casings).forEach((casing) => {
      const transformLabel = getTransformLabel(casing);

      // Now loop over each style object
      variant[casing].forEach((setting: FontSetting) => {
        const sizeLabel = `s${ parseSize(setting.fontSize) }`;
        const weightLabel = `${ styleLabel }${ transformLabel }`;

        acc[sizeLabel] = acc[sizeLabel] || { };
        acc[sizeLabel][weightLabel] = {
          fontFamily,
          fontStyle,
          fontWeight,
          fontSize: px(setting.fontSize),
          letterSpacing: setting.letterSpacing && px(setting.letterSpacing),
          lineHeight: setting.lineHeight && px(setting.lineHeight),
          textTransform: casing !== LetterCasing.normalcase ? casing : 'none',
          ...options.styles || { }
        };

        if (options.cssFn) {
          acc[sizeLabel][weightLabel] = options.cssFn(acc[sizeLabel][weightLabel]);
        }
      });
    });

    return acc;
  // tslint:disable-next-line: align
  }, { });

  return styles;
};
