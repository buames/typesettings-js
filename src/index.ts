/* eslint-disable no-nested-ternary */
import {
  FontFamilyProperty,
  FontFaceFontDisplayProperty,
  FontFaceFontFeatureSettingsProperty,
  FontFaceFontStretchProperty,
  FontFaceFontStyleProperty,
  FontFaceFontVariantProperty,
  FontFeatureSettingsProperty,
  FontSizeProperty,
  FontStyleProperty,
  FontWeightProperty,
  LineHeightProperty,
  LetterSpacingProperty,
  Properties,
} from 'csstype';

export type StyledValue = string | number | undefined | null;

export enum FontCasingTypes {
    normalcase = 'normalcase',
    uppercase = 'uppercase',
    lowercase = 'lowercase'
  }

export enum FontSourceTypes {
    locals = 'locals',
    woff2 = 'woff2',
    woff = 'woff',
    ttf = 'ttf',
    otf = 'opentype',
    eot = 'embedded-opentype'
  }

export interface Typesettings {
  family: FontFamilyProperty;
  fallbacks?: FontFamilyProperty[];
  variants: FontVariant[];
}

export interface FontVariant {
  fontStyle: FontStyleProperty;
  fontWeight: FontWeightProperty;
  sources: FontSources;
  normalcase?: FontSetting[];
  uppercase?: FontSetting[];
  lowercase?: FontSetting[];
}

export type FontSources = {
    [K in keyof typeof FontSourceTypes]?: string | string[] | NodeRequireFunction | false
  };

export interface FontSetting extends FontStyles {
  fontSize: FontSizeProperty<StyledValue>;
  letterSpacing?: LetterSpacingProperty<StyledValue>;
  lineHeight?: LineHeightProperty<StyledValue>;
}

export interface TypesettingOptions<A = CssFn, B = FontStyles, C = FontFaceStyles> {
  [k: string]: any;
  cssFn?: A;
  fontStyles?: B;
  fontFaceStyles?: C;
}

export type CssFn<T = (style: any) => any> = T;

export interface FontStyles extends Properties<StyledValue> {}

export interface FontFaceStyles extends FontStyles {
  fontDisplay?: FontFaceFontDisplayProperty;
  fontFeatureSettings?: FontFaceFontFeatureSettingsProperty;
  fontStretch?: FontFaceFontStretchProperty;
  fontStyle?: FontFaceFontStyleProperty;
  fontVariant?: FontFaceFontVariantProperty;
  fontVariationSettings?: FontFaceFontVariantProperty;
  MozFontFeatureSettings?: FontFeatureSettingsProperty;
}

export type TypesettingResults = {
  [size: string]: { [weight: string]: FontStyles };
};

/*
  Returns a pixel value or the raw css value
*/
const px = (n: StyledValue) => (
  typeof n === 'number' && n !== 0 ? `${n}px` : n
);

/*
  Parses a number and unit string and returns the size
*/
const parseSize = (str: StyledValue) => {
  if (typeof str !== 'string') return str;

  const match = str.trim().match(/([\d.\-+]*)\s*(.*)/);
  return (match && match[1]) || str;
};

/*
  Returns the font stack
*/
export const getFontStack = (family: FontFamilyProperty, fallbacks?: FontFamilyProperty[]) => ([
  normalizeFamily(family),
  fallbacks && fallbacks.map(normalizeFamily),
].filter(Boolean).join(', '));

/*
  Returns a StyledValue from a given Typesettings obj and a path to the key
*/
export const getValue = <T = StyledValue>(typesettings: Typesettings, ...paths: any[]): T => (
  paths.reduce((_, path) => path.split('.').reduce((acc: any, key: string) => (acc && acc[key] ? acc[key] : null), typesettings), null)
);

/*
  Returns a normalized FontFamily name where names with
  a space are automatically wrapped in quotes
*/
export const normalizeFamily = (family: FontFamilyProperty) => (
  /\s/g.test(family) ? `'${family}'` : family
);

/*
  Generates a map of typesettings
*/
export const generateFonts = (
  typesettings: Typesettings,
  options: TypesettingOptions = { },
): TypesettingResults => {
  /*
    Returns a weight property labelâ€š prefixed with 'n' for normal, 'i' for italics,
    and 'o' for oblique. For example, n700 equals 'normal' with a weight of '700'.
    However, if the weight is a string (ie 'bold'), this returns the prefix + a capitalized
    weight. For example, nBold.
  */
  const getStyleLabel = ({ fontStyle, fontWeight }: FontVariant) => (
    `${fontStyle.charAt(0)}${typeof fontWeight === 'string'
      ? `${fontWeight.charAt(0).toUpperCase()}${fontWeight.slice(1)}`
      : fontWeight}`
  );

  /*
  Returns a property label to append to the style property
  label depending on the lettercasing type
  */
  const getTransformLabel = (casing: FontCasingTypes | string) => (
    (casing === FontCasingTypes.uppercase) ? '_caps'
      : (casing === FontCasingTypes.lowercase) ? '_lower'
        : ''
  );

  const { family, fallbacks, variants } = typesettings;
  const fontFamily = getFontStack(family, fallbacks);

  const styles = variants.reduce((acc, variant) => {
    const {
      fontStyle, fontWeight, sources, ...casings
    } = variant;
    const styleLabel = getStyleLabel(variant);

    // Loop over each casing (ie. normalcase, uppercase, lowercase)
    Object.keys(casings).forEach((casing) => {
      const transformLabel = getTransformLabel(casing);

      // Now loop over each style object
      variant[casing].forEach((setting: FontSetting) => {
        const sizeLabel = `s${parseSize(setting.fontSize)}`;
        const weightLabel = `${styleLabel}${transformLabel}`;

        acc[sizeLabel] = acc[sizeLabel] || { };
        acc[sizeLabel][weightLabel] = {
          fontFamily,
          fontStyle,
          fontWeight,
          fontSize: px(setting.fontSize),
          letterSpacing: setting.letterSpacing && px(setting.letterSpacing),
          lineHeight: setting.lineHeight ? px(setting.lineHeight) : 'initial',
          textTransform: casing !== FontCasingTypes.normalcase ? casing : 'none',
          ...options.fontStyles || { },
        };

        if (options.cssFn) {
          acc[sizeLabel][weightLabel] = options.cssFn(acc[sizeLabel][weightLabel]);
        }
      });
    });

    return acc;
  }, { });

  return styles;
};

/*
  Generates a fontFace declaration from a typesettings obj
*/
export const generateFontFace = (
  typesettings: Typesettings,
  options: TypesettingOptions = { },
): string => {
  const { family, variants } = typesettings;

  const declaration = variants.map(({ sources, fontStyle, fontWeight }) => {
    const srcs = Object.keys(sources).map((key) => (
      sources[key] && (Array.isArray(sources[key])
        ? sources[key].map((n: string) => (`local('${n}')`))
        : `url(${sources[key]}) format('${FontSourceTypes[key]}')`)
    )).filter(Boolean);

    const styles = {
      fontStyle,
      fontWeight,
      fontFamily: normalizeFamily(family),
      src: srcs.join(', '),
      ...options.fontFaceStyles,
    };

    const face = Object.entries(styles).reduce((acc, [key, value]) => {
      const propName = key.replace(/([A-Z])/g, ((match) => `-${match[0].toLowerCase()}`));
      return `${acc}${propName}: ${value};`;
    }, '');

    return `@font-face { ${face} }`;
  });

  const fontFace = declaration.join(' ');
  return options.cssFn ? options.cssFn(fontFace) : fontFace;
};

/*
  Convenience func
*/
export const generate = (
  typesettings: Typesettings,
  options?: TypesettingOptions,
): { fontFace: string; fonts: TypesettingResults } => ({
  fontFace: generateFontFace(typesettings, options),
  fonts: generateFonts(typesettings, options),
});
