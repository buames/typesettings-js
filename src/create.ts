import { Typesettings, TypesettingOptions } from './types';
import { createFontFace } from './fontFace';
import { createFonts } from './fonts';

/**
 * Convenience function to create font styles and a @font-face declaration
 * */
export const create = <T>(
  typesettings: Typesettings,
  options?: TypesettingOptions<T>,
) => ({
  fontFace: createFontFace<T>(typesettings, options),
  fonts: createFonts<T>(typesettings, options),
});
