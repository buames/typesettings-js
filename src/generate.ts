import { Typesettings, TypesettingOptions, TypesettingResults } from './types';
import { generateFontFace } from './fontFace';
import { generateFonts } from './fonts';

export const generate = (
  typesettings: Typesettings,
  options?: TypesettingOptions
): { fontFace: string, fonts: TypesettingResults } => ({
  fontFace: generateFontFace(typesettings, options),
  fonts: generateFonts(typesettings, options)
});
