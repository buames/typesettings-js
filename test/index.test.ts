import { build } from './fixtures/a';
import { createFontFace, createFonts } from '../src';

it('returns @font-face declaration as a string', () => {
  const a = createFontFace(build());
  expect(typeof a === 'string').toBe(true);
});

it('returns @font-face declaration as snapshots', () => {
  const a = createFontFace(build({ family: 'Helvetica' }));
  expect(a).toMatchSnapshot(
    '@font-face declaration without the family wrapped in quotes',
  );

  const b = createFontFace(build({ family: 'Helvetica Neue' }));
  expect(b).toMatchSnapshot(
    '@font-face declaration with the family wrapped in quotes',
  );

  const c = createFontFace(build(), {
    fontFaceStyles: {
      fontDisplay: 'swap',
      MozFontFeatureSettings: '"tnum", "liga"',
    },
  });
  expect(c).toMatchSnapshot(
    '@font-face declaration with additional fontFace styles',
  );
});

it('returns font styles as an object', () => {
  const a = createFonts(build());
  expect(typeof a === 'object').toBe(true);
});

it('returns font styles as snapshots', () => {
  const a = createFonts(build());
  expect(a).toMatchSnapshot('Font styles as objects with fallback families');

  const b = createFontFace(build({ fallbacks: undefined }));
  expect(b).toMatchSnapshot('Font styles as objects without fallback families');

  const c = createFonts(build(), {
    fontStyles: { textRendering: 'optimizeLegibility' },
  });
  expect(c).toMatchSnapshot('Font styles with additional font styles');
});

it('returns font-family without quotes', () => {
  const a = createFonts(build());
  expect(a.s20.n500.fontFamily).toBe(
    'Helvetica, -apple-system, BlinkMacSystemFont',
  );
});

it('returns font-family without a fallback', () => {
  const a = createFonts(build({ fallbacks: undefined }));
  expect(a.s20.n500.fontFamily).toBe('Helvetica');
});

it('returns font-family with quotes', () => {
  const a = createFonts(build({ family: 'Helvetica Neue' }));
  expect(a.s20.n500.fontFamily).toBe(
    "'Helvetica Neue', -apple-system, BlinkMacSystemFont",
  );
});

it('returns font-style as italic', () => {
  const a = createFonts(build());
  expect(a.s14.i400.fontStyle).toBe('italic');
});

it('returns font-style as normal', () => {
  const a = createFonts(build());
  expect(a.s14.n400.fontStyle).toBe('normal');
});

it('returns font-size in px', () => {
  const a = createFonts(build());
  expect(a.s20.n500.fontSize).toBe('20px');
});

it('returns font-size in rem', () => {
  const a = createFonts(build());
  expect(a.s20.nBold.fontSize).toBe('20rem');
});

it('returns font-weight as a number', () => {
  const a = createFonts(build());
  expect(a.s20.n500.fontWeight).toBe(500);
});

it('returns font-weight as a string', () => {
  const a = createFonts(build());
  expect(a.s20.nBold.fontWeight).toBe('bold');
});

it('returns letter-spacing as initial', () => {
  const a = createFonts(build());
  expect(a.s14.n400.letterSpacing).toBe('initial');
});

it('returns letter-spacing in px', () => {
  const a = createFonts(build());
  expect(a.s20.n500.letterSpacing).toBe('0.29px');
});

it('returns letter-spacing in em', () => {
  const a = createFonts(build());
  expect(a.s20.nBold.letterSpacing).toBe('0.29em');
});

it('returns line-height in px', () => {
  const a = createFonts(build());
  expect(a.s20.n500.lineHeight).toBe('22px');
});

it('returns line-height in em', () => {
  const a = createFonts(build());
  expect(a.s20.nBold.lineHeight).toBe('22em');
});

it('returns line-height as normal', () => {
  const a = createFonts(build());
  expect(a.s20.n500_lower.lineHeight).toBe('normal');
});

it('returns text-transform as undefined', () => {
  const a = createFonts(build());
  expect(a.s20.n500.textTransform).toBeUndefined();
});

it('returns text-transform as uppercase', () => {
  const a = createFonts(build());
  expect(a.s20.n500_caps.textTransform).toBe('uppercase');
});

it('returns text-transform as lowercase', () => {
  const a = createFonts(build());
  expect(a.s20.n500_lower.textTransform).toBe('lowercase');
});
