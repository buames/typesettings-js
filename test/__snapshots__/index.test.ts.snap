// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns @font-face declaration as snapshots: @font-face declaration with additional fontFace styles 1`] = `"@font-face { font-style: italic;font-weight: 400;font-family: Helvetica;src: local('Helvetica Regular'),local('Helvetica-Regular');font-display: swap;-moz-font-feature-settings: \\"tnum\\", \\"liga\\"; } @font-face { font-style: normal;font-weight: 400;font-family: Helvetica;src: local('Helvetica Regular'),local('Helvetica-Regular');font-display: swap;-moz-font-feature-settings: \\"tnum\\", \\"liga\\"; } @font-face { font-style: normal;font-weight: 500;font-family: Helvetica;src: local('Helvetica Medium'),local('Helvetica-Medium'), url(./font-file.eot?#iefix) format('embedded-opentype'), url(./font-file.woff) format('woff'), url(./font-file.woff2) format('woff2'), url(./font-file.ttf) format('ttf');font-display: swap;-moz-font-feature-settings: \\"tnum\\", \\"liga\\"; } @font-face { font-style: normal;font-weight: bold;font-family: Helvetica;src: local('Helvetica Bold'),local('Helvetica-Bold'), url(./font-file.eot?#iefix) format('embedded-opentype'), url(./font-file.woff) format('woff'), url(./font-file.woff2) format('woff2'), url(./font-file.ttf) format('ttf');font-display: swap;-moz-font-feature-settings: \\"tnum\\", \\"liga\\"; }"`;

exports[`returns @font-face declaration as snapshots: @font-face declaration with the family wrapped in quotes 1`] = `"@font-face { font-style: italic;font-weight: 400;font-family: 'Helvetica Neue';src: local('Helvetica Regular'),local('Helvetica-Regular'); } @font-face { font-style: normal;font-weight: 400;font-family: 'Helvetica Neue';src: local('Helvetica Regular'),local('Helvetica-Regular'); } @font-face { font-style: normal;font-weight: 500;font-family: 'Helvetica Neue';src: local('Helvetica Medium'),local('Helvetica-Medium'), url(./font-file.eot?#iefix) format('embedded-opentype'), url(./font-file.woff) format('woff'), url(./font-file.woff2) format('woff2'), url(./font-file.ttf) format('ttf'); } @font-face { font-style: normal;font-weight: bold;font-family: 'Helvetica Neue';src: local('Helvetica Bold'),local('Helvetica-Bold'), url(./font-file.eot?#iefix) format('embedded-opentype'), url(./font-file.woff) format('woff'), url(./font-file.woff2) format('woff2'), url(./font-file.ttf) format('ttf'); }"`;

exports[`returns @font-face declaration as snapshots: @font-face declaration without the family wrapped in quotes 1`] = `"@font-face { font-style: italic;font-weight: 400;font-family: Helvetica;src: local('Helvetica Regular'),local('Helvetica-Regular'); } @font-face { font-style: normal;font-weight: 400;font-family: Helvetica;src: local('Helvetica Regular'),local('Helvetica-Regular'); } @font-face { font-style: normal;font-weight: 500;font-family: Helvetica;src: local('Helvetica Medium'),local('Helvetica-Medium'), url(./font-file.eot?#iefix) format('embedded-opentype'), url(./font-file.woff) format('woff'), url(./font-file.woff2) format('woff2'), url(./font-file.ttf) format('ttf'); } @font-face { font-style: normal;font-weight: bold;font-family: Helvetica;src: local('Helvetica Bold'),local('Helvetica-Bold'), url(./font-file.eot?#iefix) format('embedded-opentype'), url(./font-file.woff) format('woff'), url(./font-file.woff2) format('woff2'), url(./font-file.ttf) format('ttf'); }"`;

exports[`returns font styles as snapshots: Font styles as objects with fallback families 1`] = `
Object {
  "s14": Object {
    "i400": Object {
      "fontFamily": "Helvetica, -apple-system, BlinkMacSystemFont",
      "fontSize": "14px",
      "fontStyle": "italic",
      "fontWeight": 400,
      "letterSpacing": "initial",
      "lineHeight": "18px",
    },
    "n400": Object {
      "fontFamily": "Helvetica, -apple-system, BlinkMacSystemFont",
      "fontSize": "14px",
      "fontStyle": "normal",
      "fontWeight": 400,
      "letterSpacing": "initial",
      "lineHeight": "18px",
    },
  },
  "s20": Object {
    "n500": Object {
      "fontFamily": "Helvetica, -apple-system, BlinkMacSystemFont",
      "fontSize": "20px",
      "fontStyle": "normal",
      "fontWeight": 500,
      "letterSpacing": "0.29px",
      "lineHeight": "22px",
    },
    "n500_caps": Object {
      "fontFamily": "Helvetica, -apple-system, BlinkMacSystemFont",
      "fontSize": "20px",
      "fontStyle": "normal",
      "fontWeight": 500,
      "letterSpacing": "1px",
      "lineHeight": "1px",
      "textTransform": "uppercase",
    },
    "n500_lower": Object {
      "fontFamily": "Helvetica, -apple-system, BlinkMacSystemFont",
      "fontSize": "20px",
      "fontStyle": "normal",
      "fontWeight": 500,
      "letterSpacing": "initial",
      "lineHeight": "normal",
      "textTransform": "lowercase",
    },
    "nBold": Object {
      "fontFamily": "Helvetica, -apple-system, BlinkMacSystemFont",
      "fontSize": "20rem",
      "fontStyle": "normal",
      "fontWeight": "bold",
      "letterSpacing": "0.29em",
      "lineHeight": "22em",
    },
  },
  "sFooBar": Object {
    "nBold": Object {
      "fontFamily": "Helvetica, -apple-system, BlinkMacSystemFont",
      "fontSize": "FooBar",
      "fontStyle": "normal",
      "fontWeight": "bold",
      "letterSpacing": "0.29em",
      "lineHeight": "22em",
    },
  },
}
`;

exports[`returns font styles as snapshots: Font styles as objects without fallback families 1`] = `"@font-face { font-style: italic;font-weight: 400;font-family: Helvetica;src: local('Helvetica Regular'),local('Helvetica-Regular'); } @font-face { font-style: normal;font-weight: 400;font-family: Helvetica;src: local('Helvetica Regular'),local('Helvetica-Regular'); } @font-face { font-style: normal;font-weight: 500;font-family: Helvetica;src: local('Helvetica Medium'),local('Helvetica-Medium'), url(./font-file.eot?#iefix) format('embedded-opentype'), url(./font-file.woff) format('woff'), url(./font-file.woff2) format('woff2'), url(./font-file.ttf) format('ttf'); } @font-face { font-style: normal;font-weight: bold;font-family: Helvetica;src: local('Helvetica Bold'),local('Helvetica-Bold'), url(./font-file.eot?#iefix) format('embedded-opentype'), url(./font-file.woff) format('woff'), url(./font-file.woff2) format('woff2'), url(./font-file.ttf) format('ttf'); }"`;

exports[`returns font styles as snapshots: Font styles with additional font styles 1`] = `"@font-face { font-style: italic;font-weight: 400;font-family: Helvetica;src: local('Helvetica Regular'),local('Helvetica-Regular'); } @font-face { font-style: normal;font-weight: 400;font-family: Helvetica;src: local('Helvetica Regular'),local('Helvetica-Regular'); } @font-face { font-style: normal;font-weight: 500;font-family: Helvetica;src: local('Helvetica Medium'),local('Helvetica-Medium'), url(./font-file.eot?#iefix) format('embedded-opentype'), url(./font-file.woff) format('woff'), url(./font-file.woff2) format('woff2'), url(./font-file.ttf) format('ttf'); } @font-face { font-style: normal;font-weight: bold;font-family: Helvetica;src: local('Helvetica Bold'),local('Helvetica-Bold'), url(./font-file.eot?#iefix) format('embedded-opentype'), url(./font-file.woff) format('woff'), url(./font-file.woff2) format('woff2'), url(./font-file.ttf) format('ttf'); }"`;
